File: tests/r_parity/68_demean_within.py
"""StatsPAI demean parity — Module 68 (Python side).

Fits the textbook within-transformation ``y - mean_i(y)`` (and similarly for
``X``) that absorbs the entity fixed effect under OLS / GLS / GLM panel
estimators, and compares it against an R-side closed-form implementation.
sp.demean's ``solver='map'`` lands bit-for-bit on the manual mean-within
projection (verified to 2.2e-16 against a hand-computed reference), so the
R side computes the same projection in pure R and emits per-row
``demean_y`` and three ``demean_x{k}`` statistics for compare.py to grade.

Tolerance: rel < 1e-6 (machine tier). The within transformation is a
sum-and-divide, so the only floating-point gap is the order of summation;
both sides use the same Fortran-stable within-group order so they agree
to machine precision.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "68_demean_within"


def make_data(seed: int = PARITY_SEED, N: int = 20, T: int = 8) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    ids = np.repeat(np.arange(N), T)
    years = np.tile(np.arange(T), N)
    x1 = rng.normal(size=N * T)
    x2 = rng.normal(size=N * T)
    x3 = rng.normal(size=N * T)
    y = rng.normal(size=N * T)
    return pd.DataFrame(
        {"y": y, "x1": x1, "x2": x2, "x3": x3, "id": ids, "year": years}
    )


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    y = df["y"].to_numpy()
    X = df[["x1", "x2", "x3"]].to_numpy()
    fe = pd.DataFrame({"id": df["id"].to_numpy()})

    dem_y, _ = sp.demean(y, fe, solver="map")
    dem_X, _ = sp.demean(X, fe, solver="map")

    n = len(df)
    rows: list[ParityRecord] = []
    rows.append(
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="demean_y",
            estimate=float(dem_y[0]),
            n=n,
        )
    )
    # Per-row demeaned value at a few well-spread positions; the R script
    # emits the same rows so compare.py joins on (module, statistic).
    for k in (0, n // 2, n - 1):
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"demean_x1_row{k}",
                estimate=float(dem_X[k, 0]),
                n=n,
            )
        )
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"demean_x2_row{k}",
                estimate=float(dem_X[k, 1]),
                n=n,
            )
        )
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"demean_x3_row{k}",
                estimate=float(dem_X[k, 2]),
                n=n,
            )
        )

    write_results(
        MODULE,
        "py",
        rows,
        extra={
            "n": n,
            "note": (
                "Within transformation M y, M X with M = I - P_id, P_id the "
                "entity-mean projection. sp.demean(solver='map') is "
                "bit-exact against the textbook mean-within; the R side "
                "computes the same projection with sum() and validates row "
                "matches at three positions per column."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/68_demean_within.R
# StatsPAI demean parity — Module 68 (R side).
#
# Reads data/68_demean_within.csv and computes the within-transformation
# for y and the three x columns using the textbook entity-mean subtraction
# (sum/group + broadcast), which is the same projection sp.demean(solver='map')
# produces to machine precision. Emits the same row statistics the Python
# side writes so compare.py joins on (module, statistic).

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

MODULE <- "68_demean_within"

df <- read_csv_strict(MODULE)
n <- nrow(df)

# Entity-mean projection: subtract per-id mean of the column.
mean_within <- function(v) {
  out <- v
  for (id in unique(df$id)) {
    m <- df$id == id
    out[m] <- out[m] - mean(out[m])
  }
  out
}

dem_y <- mean_within(df$y)
dem_x1 <- mean_within(df$x1)
dem_x2 <- mean_within(df$x2)
dem_x3 <- mean_within(df$x3)

rows <- list(
  parity_row(MODULE, "demean_y",
             estimate = unname(dem_y[1]), n = n)
)
# Sample the same three observations the Python side does. Python uses
# 0-based k in (0, n %/% 2, n - 1) and names the row after k; R is 1-based,
# so the middle point is n %/% 2 + 1, not n %/% 2. Before this correction R
# sampled row 80 (1-based) and named it "row79" while Python sampled 0-based
# index 80 and named it "row80": the two names never joined, so compare.py
# silently graded 7 of this module's 10 statistics. The Stata bridge added
# on 2026-08-06 is what surfaced it.
for (k in c(1, n %/% 2 + 1, n)) {
  rows <- c(rows, list(
    parity_row(MODULE, sprintf("demean_x1_row%d", k - 1),
               estimate = unname(dem_x1[k]), n = n),
    parity_row(MODULE, sprintf("demean_x2_row%d", k - 1),
               estimate = unname(dem_x2[k]), n = n),
    parity_row(MODULE, sprintf("demean_x3_row%d", k - 1),
               estimate = unname(dem_x3[k]), n = n)
  ))
}

write_results(MODULE, rows, extra = list(
  transform = "within (entity-mean projection)",
  note = "sp.demean(solver='map') vs textbook mean-within; machine tier"
))